src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,158 @@
+* a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction,
+* including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software,
+* and to permit persons to whom the Software is furnished to do so,
+* subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+#include "util/math.h"
+#include "util/sample.h"
+
+using ReadableSlice = mixxx::SampleBuffer::ReadableSlice;
+using WritableSlice = mixxx::SampleBuffer::WritableSlice;
+
+namespace {
+SINT copyRing(const ReadableSlice pSourceBuffer,
+        SINT sourcePos,
+        const WritableSlice pDestBuffer,
+        SINT destPos,
+        const SINT numItems) {
+    const SINT newSourcePos = sourcePos + numItems;
+    const SINT newDestPos = destPos + numItems;
+
+    VERIFY_OR_DEBUG_ASSERT(newSourcePos <= pSourceBuffer.length() ||
+            newDestPos <= pDestBuffer.length()) {
+        return 0;
+    }
+
+    // Check to see if the copy is not contiguous.
+    if (newSourcePos > pSourceBuffer.length() ||
+            newDestPos > pDestBuffer.length()) {
+        // Copy is not contiguous.
+        SINT firstDataBlockSize = math_min(pSourceBuffer.length() - sourcePos,
+                pDestBuffer.length() - destPos);
+
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                firstDataBlockSize);
+
+        sourcePos = (sourcePos + firstDataBlockSize) % pSourceBuffer.length();
+        destPos = (destPos + firstDataBlockSize) % pDestBuffer.length();
+
+        // The second data part is the start of the ring buffer.
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                numItems - firstDataBlockSize);
```

Comment:
I think this breaks if both buffers would requires non-contigous reads. 
```
copying 6 items of a buffer of length 8:
input:
0000|00 00xxxx
1111 11|11xxxx
after first copy:
 000000|00xxxx
|111111 00xxxx
after second copy:
0000 0000xx|xx
00xx|1100
``` 
xx is out-of-bounds data/memory

Reply:
That is why we asset that is does not happen above.

Reply:
Ah yes, I misinterpreted the code above. That makes sense now.